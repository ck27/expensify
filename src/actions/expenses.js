import {v4 as uuidV4} from "uuid";

export const addExpense = ( { description = "", note = "", amount= 0, createdAt = 0 } = {}) => ({
    type: "ADD_EXPENSE",
    expense: {
        id: uuidV4(),
        description,
        note,
        amount,
        createdAt
    }
});

export const deleteExpense = ( { id } = {} ) => ( {
    type:"DELETE_EXPENSE",
    id
});

export const editExpense = ( expense, updates ) => ({
    type: "EDIT_EXPENSE",
    expense,
    updates
});