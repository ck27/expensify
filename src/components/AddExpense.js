import React from "react";

const AddExpense = () => (
    <div>
        <h1>Add Expense</h1>
    </div>
);

export default AddExpense;