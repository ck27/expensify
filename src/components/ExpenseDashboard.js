import React from "react";
import { Link } from "react-router-dom";
import ExpenseList from "./ExpenseList";

const ExpenseDashboard = () => (
    <div>
        <h1>Expense Dashboard</h1>

        <Link to="/edit/1">Expense : 1</Link>
        <ExpenseList />
    </div>
);

export default ExpenseDashboard;