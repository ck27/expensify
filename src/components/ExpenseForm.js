import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import "react-dates/initialize";


const now = moment();

export default class ExpenseForm extends React.Component {

    constructor(props) {
        super(props); 
        this.state = {
            description: props.expense ? props.expense.description : "",
            note: props.expense ? props.expense.note : "",
            amount: props.expense ? props.expense.amount : "",
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            isValid: true
        };
    }


    onDescChange = (e) => {
        const description = e.target.value;
        
        this.setState( () => ({ description}));
        console.log(this.state);
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState( () => ({ note }));
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        if( !amount || amount.match(/^\d+(\.\d{0,2})?$/) ) {
            this.setState( () => ({ amount }) );
        }
    }

    onDateChange = (createdAt) => {
        if(createdAt) {
            this.setState( () => ({createdAt}))
        }
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({
            calendarFocused : focused   
        }));
    }

    onSubmit = (e) => {
        console.log('Expense Form submitted')
        e.preventDefault();
        if(!this.state.description || !this.state.amount) {
            this.setState(() => ({isValid :false}))
        } else {
            this.setState(() => ({isValid :true}));
            console.log("submitted dispatch");
            this.props.onSubmit({
                description: this.state.description,
                amount : parseFloat(this.state.amount,10),
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    }

    render() {
        return (
            <div>
                {!this.state.isValid && <p>Please enter valid input for desription amount</p>}
                <form onSubmit={this.onSubmit}>
                    <input type="text" placeholder="Description" value={this.state.description} onChange={this.onDescChange} />
                    <br/>
                    <input type="text" placeholder="Amount" value={this.state.amount} onChange={this.onAmountChange}/>
                    <br/>
                    <textarea placeholder="Notes for your expense (optional)" value={this.state.note} onChange={this.onNoteChange} />
                    <br/>
                    <SingleDatePicker 
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        id="createdAt"
                        numberOfMonths={1}
                        isOutsideRange={ () => ( false )}
                    />
                    <br/>
                    <br/>
                    <button>Add Expense</button>
                </form>
            </div>
        );
    }
}