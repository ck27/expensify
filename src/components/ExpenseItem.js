import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";

const ExpenseItem = ({ id, description, amount, createdAt, dispatch}) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <div>
            <p>{amount} Paid at {createdAt}</p>
        </div>
    </div>
);

export default ExpenseItem;