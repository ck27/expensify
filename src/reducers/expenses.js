const expensesReducerDefaultState = [];

export const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch(action.type){
        case "ADD_EXPENSE":
            return [
                ...state,
                action.expense
            ];
        case "DELETE_EXPENSE":
            return state.filter( ( {id} ) => id !== action.expenseId );

        case "EDIT_EXPENSE":
            return state.map( (exp) => {
                if(exp.id == action.expense.id) {
                    return {
                        ...exp,
                        ...action.expense
                    };
                } else {
                    return exp;
                }
            });
        default:
            return state;
    }

};
