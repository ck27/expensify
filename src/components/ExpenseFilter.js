import React from 'react';
import {connect} from "react-redux";
import {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} from "../actions/filters";
import { DateRangePicker } from 'react-dates';


export class ExpenseFilter extends React.Component {

    constructor(props) {
        super(props);
    }

    state = {
        calendarFocused : null
    };

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange = (calendarFocused) => {
        this.setState( () => ({calendarFocused}) );
    }

    onTextFilterChange = (e)=> {
        this.props.setTextFilter(e.target.value);
    };

    onSortFilterChange = (e) => {
        if(e.target.value == "date") {
            return this.props.sortByDate();
        }
        if(e.target.value == "amount"){
            return this.props.sortByAmount();
        } 
    };

    render() {
        return (
            <div>
                <input type="text" value={this.props.filter.text} onChange={this.onTextFilterChange}/>
                <select value={this.props.filter.text} onChange={this.onSortFilterChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filter.startDate}
                    startDateId="2020-06-01"    
                    endDate={this.props.filter.endDate}
                    endDateId="2020-06-30"
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}    
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
            </div>
        );
    }
}

const mapStateToProps = (state) => (
    {
        filter: state.filter
    }
);

const mapDispatchToProps = (dispatch) => {
    return {
        sortByDate : () => dispatch(sortByDate),
        sortByAmount: () => dispatch(sortByAmount),
        setTextFilter: (text) => dispatch(setTextFilter(text)),
        setStartDate: (date) => dispatch(setStartDate(date)),
        setEndDate: (date) => dispatch(setEndDate(date))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseFilter);