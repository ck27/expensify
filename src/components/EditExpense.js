import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { editExpense, deleteExpense } from "../actions/expenses";

export class EditExpense extends React.Component {
    onSubmit = (expense) => {
        console.log(expense);
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push("/");
    }

    onDelete = () => {
        this.props.deleteExpense(this.props.expense.id);
        this.props.history.push("/");
    }

    render() {
        return (
            <div>
                <h1>Edit Expense : {this.props.match.params.id}</h1>
                <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit}/>
                <button onClick={this.onDelete}>REMOVE</button>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find( (exp) => exp.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch,props) => ({
    editExpense : (id, expense) => dispatch(editExpense(id, expense)),
    deleteExpense : (id) => dispatch(deleteExpense(id))    
})

export default connect(mapStateToProps,mapDispatchToProps)(EditExpense);