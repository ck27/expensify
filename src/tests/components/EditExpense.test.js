import React from "react";
import {EditExpense} from "../../components/EditExpense";
import { shallow } from "enzyme";
import expenses from "../fixtures/expenses";

let editExpense, deleteExpense, history, content, anExpense;

beforeEach( () => {
    editExpense = jest.fn();
    deleteExpense = jest.fn();
    history = { push : jest.fn() };

    anExpense = expenses[2];
    const match =  {
        'params' : {
            'id' : anExpense.id
        }
    };
    content = shallow(<EditExpense expense={anExpense} editExpense={editExpense} deleteExpense={deleteExpense} history={history}  match={match}/>);
});

test("it should render EditExpense", () => {
    expect(content).toMatchSnapshot();
});

test("it should be able to edit expense", () => {
    content.find("ExpenseForm").prop("onSubmit")(anExpense);
    expect(editExpense).toHaveBeenLastCalledWith(anExpense.id, anExpense);
    expect(history.push).toHaveBeenLastCalledWith("/");
});

test("it should be able to delete expense", () => {
    content.find("button").simulate("click");
    expect(history.push).toHaveBeenLastCalledWith("/");
    expect(deleteExpense).toHaveBeenLastCalledWith(anExpense.id);
});