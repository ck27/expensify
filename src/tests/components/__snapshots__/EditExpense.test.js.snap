// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should render EditExpense 1`] = `
<div>
  <h1>
    Edit Expense : 
    3
  </h1>
  <ExpenseForm
    expense={
      Object {
        "amount": 3500,
        "createdAt": 345600000,
        "description": "BESCOM Bill",
        "id": 3,
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    REMOVE
  </button>
</div>
`;
